### PR TITLE
Upgrade to ASM 6

### PR DESCRIPTION
**Pull request by melix**
### Context

Our Java 9 build is still not ready, however, there are some tests which are disabled today because mocking is broken with `cglib` and `asm5`. A simple way to re-enable those tests are recover coverage is to upgrade to ASM 6 alpha. It's worth noting that the alpha here mostly states for the newer parts of the API, that is to say modules, so shouldn't impact the rest of the codebase.

ASM 6 final is only going to be released _after_ the final release of JDK 9, so we need to do something before in any case: either using alpha/beta versions, or disable JDK 9 tests.

This commit doesn't consider what to do, in practice, when we find a `module-info.class` file on classpath wrt to ABI (compile avoidance). This should be solved in a different PR (if needs be).

### CODE REVIEW

File: subprojects/testing-base/src/test/groovy/org/gradle/api/internal/tasks/testing/logging/GroovyStackTraceSpecTest.groovy [hunk 1]
```diff
@@ -74,4 +74,12 @@ class GroovyStackTraceSpecTest extends Specification {
             throw new Exception("boom")
         }
     }
+
+    private static void stackTraceMatches(StackTraceElement element, String className, String methodName, String fileName, int lineNumber) {
```
Comment (eljobe) [inline subprojects/testing-base/src/test/groovy/org/gradle/api/internal/tasks/testing/logging/GroovyStackTraceSpecTest.groovy]: ⭕️ I don't really understand why this change is part of this commit.

Comment (melix) [inline subprojects/testing-base/src/test/groovy/org/gradle/api/internal/tasks/testing/logging/GroovyStackTraceSpecTest.groovy]: Because `StackTraceElement#equals` doesn't work anymore in JDK 9 : it also compares the class loaders, for example.

**Comment by adammurdoch** [comment]
I want to change "approve" to "request changes": make sure `Java9ClassReader` is deleted as we should no longer require this.

**Comment by melix** [comment]
Good catch @adammurdoch , I forgot to remove it. It's done now.
